fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/AbstractDataAccessRule.java

```diff
@@ -227,31 +232,187 @@ protected Set<? extends RelationalExpression> dataAccessForMatchPartition(@Nonnu
+        //
+        final BitSet[] sieveBitMatrix =
+                bestMaximumCoverageMatches.size() > 2 ? newSquareBitMatrix(bestMaximumCoverageMatches.size()) : null;
+        for (final var binaryPartition : ChooseK.chooseK(bestMaximumCoverageMatches, 2)) {
+            Verify.verify(binaryPartition.size() == 2);
+            final var binaryIntersections =
+                    createIntersectionAndCompensation(
+                            call,
+                            commonPrimaryKeyValues,
+                            bestMatchToDistinctPlanMap,
+                            binaryPartition,
+                            requestedOrderings);
+            if (binaryIntersections.hasCommonIntersectionOrdering()) {
+                resultSet.addAll(binaryIntersections.getExpressions());
+            } else {
+                if (sieveBitMatrix != null) {
+                    final int i = binaryPartition.get(0).getPosition();
+                    final int j = binaryPartition.get(1).getPosition();
+                    sieveBitMatrix[i].set(j);
+                    sieveBitMatrix[j].set(i);
+                }
+            }
+        }
+
+        if (bestMaximumCoverageMatches.size() > 2) {
+            Objects.requireNonNull(sieveBitMatrix);
+
+            final var checkBitMatrix = newSquareBitMatrix(bestMaximumCoverageMatches.size());
+
+            int numCombinations = 0;
+            int numDiscardedCombinations = 0;
+
+            // create intersections for all n choose k partitions from k = 3 ... n
+            for (int k = 3; k < bestMaximumCoverageMatches.size() + 1; k ++) {
+                boolean hasCommonOrderingForK = false;
+                for (final var kPartition : ChooseK.chooseK(bestMaximumCoverageMatches, k)) {
+                    numCombinations ++;
+                    call.emitEvent(Debugger.Location.ALL_INTERSECTION_COMBINATIONS);
+
+                    //
+                    // For a combination of n orderings we enumerate find all the positions in that the combination
+                    // does not contain. Starting from a matrix where all bits are set we clear all columns and
+                    // rows at those positions. That is the check bit matrix.
+                    //
+                    setAll(checkBitMatrix); // all bits to 1
+                    final var membership = new BitSet(bestMaximumCoverageMatches.size());
+                    for (final var singleMatchedAccessVectored : kPartition) {
+                        membership.set(singleMatchedAccessVectored.getPosition());
+                    }
+
+                    for (int position = membership.nextClearBit(0);
+                             position >= 0 && position < bestMaximumCoverageMatches.size();
+                             position = membership.nextClearBit(position + 1)) {
+                        clearRowAndColumnAtPosition(checkBitMatrix, position);
+                    }
```

Comment:
I believe we can make this slightly more efficient by getting rid of `membership` bitset:

```java
                     // setAll(checkBitMatrix); // all bits to 1, this is not needed
                     // final var membership = new BitSet(bestMaximumCoverageMatches.size());
                     for (final var singleMatchedAccessVectored : kPartition) {
                         setRowAndColumnAtPosition(checkBitMatrix, position); // method that sets the row and column according to the position.
                     }
```

This works the inverse way starting off with `checkBitMatrix` set to zero, then setting respective `kPartition` positions to one. But this has the same worst-case cost as your approach, because setting 0-bits for `K:[3,N]` is _almost_ equal to setting 1-bits for `K:[3,N]`.

Reply:
After some discussion, this is not possible because we need to unset the bits of the non-participating indexes. Thanks @normen662 for the clarification.